cmd/zfs/zfs_main.c

```diff
@@ -4383,6 +4383,100 @@ zfs_do_snapshot(int argc, char **argv)
 	return (-1);
 }
 
+typedef struct zfs_send_exclude_arg {
+	size_t count;
+	char **list;
+} zfs_send_exclude_arg_t;
+
+/*
+ * This function creates the zfs_send_exclude_arg_t
+ * object described above; it can be called multiple
+ * times, and the input can be comma-separated.
+ * This is NOT the most efficient data layout; however,
+ * I couldn't think of a non-pathological case where
+ * it should have more than a couple dozen instances
+ * of excludes. If that turns out to be used in
+ * practice, we might want to instead use a tree.
+ */
+static void
+add_dataset_excludes(char *in_exclude, zfs_send_exclude_arg_t *context)
+{
+	char *exclude = strdup(in_exclude);
+	size_t tempSize = context->count;
+	char **tempArray = context->list;
+	char *tempString = NULL;
+	char *exclude_orig = exclude;
+
+	while ((tempString = strsep(&exclude, ",")) != NULL) {
+		tempArray = realloc(tempArray,
+		    (sizeof (char **)) * (tempSize + 1));
+		if (tempArray == NULL)
+			nomem();
+		tempArray[tempSize] = strdup(tempString);
+		if (tempArray[tempSize] == NULL)
+			nomem();
```

Comment:
Both of these have `safe_*()` versions.

Also, you can just strtok over `in_exclude` here:
```c
static void
add_dataset_excludes(char *exclude, zfs_send_exclude_arg_t *context)
{
	for (char *tok; (tok = strsep(&exclude, ",")) != NULL; ) {
		context->list = safe_realloc(context->list,
		    (sizeof (char *)) * (context->count + 1));
		context->list[tempSize] = tok;
		tempSize++;
	}
}

static void
free_dataset_excludes(zfs_send_exclude_arg_t *exclude_list)
{
	free(exclude_list->list);
}
```
and with that, just inline those into their single callsites.

Reply:
For strtok: "This interface is obsoleted by strsep(3)."  I started out with strtok, checked a couple of systems man pages, and deliberately decided to go with strsep as a result.

I strdup'd because I am always wary of modifying something another function gave me, and my uncertainty whether it is safe to do so.  The cost is very minor -- as I said above, I don't expect to have more than a handful of excludes unless someone is trying to be difficult, and if they are, there's a bigger cost in the layout of the array, rather than processing the entries.

I can change it to safe_realloc & safe_strdup.

Reply:
It is safe to do so, even on systems where the environment block is read-only, because main strdups argv *for precisely this reason*. Pretty much all subcommands, especially ones that decode lists and K/Vs (cf. `parseprop()`, `parsepropname()`, &c., defined locally, and the `getsubopt()` calls), modify argv – because that's what it's for.  Reallocating this is a lot of spaghetti code for no reason at all (indeed, you're effectively duplicating main() code here).

Reply:
I am not duplicating the zfs_save_arguments code at all, changing argv elements results in changing was ps shows on some systems which may matter to some people, but I am honestly just tired of arguing it so I'm just using a simplified version of your code.

Reply:
*What* are you on about. zfs_save_arguments.. saves arguments for history. I don't think I can more clearly say that *main already strdups argv* – https://github.com/openzfs/zfs/blob/a86e089415679cf1b98eb424a159bb36aa2c19e3/cmd%2Fzfs%2Fzfs_main.c#L8658-L8661

Reply:
I honestly don't know any more, but the code is the way you want, it works, and uses less memory, so that's good. ;)